[MRG+1] ENH Add working_memory global config for chunked operations

jnothman opened the pull request:
We often get issues related to memory consumption and don't deal with them particularly well.  Indeed, Scikit-learn should be at home on commodity hardware like developer/researcher laptops.

Some operations can be performed chunked, so that the result is computed in constant (or `O(n)`) memory relative to some current `O(n)` (`O(n^2)`) consumption.  Examples include: getting the argmin and min of all pairwise distances (currently done with an ad-hoc parameter to [`pairwise_distances_argmin_min`](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise_distances_argmin_min.html)), calculating silhouette score (#1976), getting nearest neighbors with brute force (#7287), calculating standard deviation of every feature (#5651). 

It's not very helpful to provide this "how much constant memory" parameter in each function (because they're often called within nested code), so this PR instead makes a global config parameter of it. The optimisation is then transparent to the user, but still configurable.

At @rth's request, this PR has been cut back.  The proposed changes to silhouette and neighbors can be seen [here](https://github.com/scikit-learn/scikit-learn/compare/master...jnothman:working_memory-plus?expand=1).

This PR (building upon my work with @dalmia) will therefore:

* add `set_config(working_memory=n_mib)`
* add `pairwise_distances_chunked`
* make use of the latter in ~nearest neighbors, silhouette and~ pairwise_distances_argmin_min
* deprecate `batch_size` in `pairwise_distances_argmin_min`

and thus:

* help towards fixing #7175, #10279, #7177 (silhouette)
* Resolve #7979
* help towards fixing #7287 (neighbors)
* perhaps help towards #8216
* provide an interface for fixing #5651
* it looks like there was some suggestion of using this kind of chunking in LMNN (#8602)

TODO:
* [x] attract other core devs' attention which I feel has been lacking despite the repeated popular interest in these issues!
* [x] review my comments to @dalmia at #7979 
* [x] add tests for `get_chunk_n_rows`
* [x] add tests for `_check_chunk_size` and any others for `pairwise_distances_chunked`
* [x] add config documentation
* [ ] fix, perhaps, #5651 with a chunked standard deviation implementation (can be a separate PR)
* [x] perhaps review uses of `gen_batches` to see if they can use this parameter to be more principled in the choice of batch size: in most cases the batch size affects the model, so we can't just change it. And in other cases, it may make little difference. We could change mean_shift's use from fixed batch_size=500 to something sensitive to working_memory if we wish.
* [x] add what's new entry
* [x] add example to `pairwise_distances_chunked` that uses start and a tuple return
* [x] benchmarking

Sentient07 committed:
Reverted the change, added regression test

reverted the comment

Resolved merge conflicts

jnothman committed:
ENH block_size for memory efficiency in silhouette

jnothman committed:
Merge remote-tracking branch 'upstream/pr/6089' into silhouette-chunks

jnothman committed:
DOC add versionadded to new parameter

jnothman committed:
FIX use bincount instead of np.add.at for old numpy

jnothman committed:
ENH use unary pairwise_distances where possible

jnothman committed:
DOC explicit block_size parameter in silhouette_score

jnothman committed:
TST test silhouette_samples explicitly

jnothman committed:
ENH support n_jobs in silhouette_score

and add comments

jnothman committed:
DOC update block_size description given n_jobs

jnothman committed:
DOC docstring formatting

jnothman committed:
ENH specify silhouette block size in bytes

jnothman committed:
block_size specified in MiB

jnothman committed:
document parameters to silhouette helper

jnothman committed:
FIX pass n_jobs from silhouette_score

Also use threading for parallelism

dalmia committed:
ENH: Added template for pairwise_distances_blockwise with docstring changes

dalmia committed:
ENH: added generator of blocks based on block_size

dalmia committed:
FIX: removed errors and extra value for metric

dalmia committed:
FIX: remove redundant variables

dalmia committed:
FIX: remove flake8 errors

dalmia committed:
BUG: added fix for Y=None

dalmia committed:
FIX: remove whitespace

dalmia committed:
FIX: fix typo

dalmia committed:
TST: added tests for pairwise_distances_blockwise

dalmia committed:
FIX: removed errors and modified pairwise_distances_blockwise with tests

dalmia committed:
FIX: fix typo

dalmia committed:
WIP: support for nearest neighbors

dalmia committed:
ENH: passing arguments to reduce_func via partial

dalmia committed:
FIX: remove true_distances as parameter

dalmia committed:
FIX: revert unintended change

dalmia committed:
FIX: convert float indices to int

dalmia committed:
FIX: removed debug lines

dalmia committed:
ENH: added pairwise_distances_reduce for radius_neighbors

dalmia committed:
FIX: changed order of reduce_func

dalmia committed:
FIX: get pairwise_distances_reduce to work correctly

dalmia committed:
FIX: remove flake8 errors

dalmia committed:
FIX: rename reduce_func

dalmia committed:
FIX: return stacked distances from pairwise_distances_reduce

dalmia committed:
TST: added tests for pairwise_distances_reduce

dalmia committed:
FIX: correct doctests

dalmia committed:
FIX: remove conflicting doctests for Python2 and Python3

dalmia committed:
FEAT: add new file for flexible_vstack

dalmia committed:
FIX: resolve conflicts on tests

dalmia committed:
FIX: remove block_size placeholders from neighbors

dalmia committed:
FIX: use generator expressions

dalmia committed:
FIX: replace error on invalid block_size with warning

dalmia committed:
FIX: replace error on invalid block_size with warning

dalmia committed:
TST: check each components meets specified memory requirement

dalmia committed:
FIX: remove PEP8 errors

dalmia committed:
ENH: move flexible_vstack to __init__

dalmia committed:
TST: add tests for flexible_vstack

dalmia committed:
DOC: improve docstring for

dalmia committed:
FIX: correct X, y for Python3

dalmia committed:
ENH: rewrote pairwise_distances_argmin_min using pairwise_distances_reduce

jnothman committed:
Merge branch 'master' into blockwise

jnothman committed:
Merge branch 'silhouette-chunks' into blockwise

jnothman committed:
[WIP] ENH Add working_memory global config for chunked operations

jnothman committed:
Add to classes.rst

jnothman committed:
Add missing module

jnothman committed:
Remove obsolete TODOs

jnothman commented on sklearn/utils/__init__.py:
```diff
@@ -506,3 +508,146 @@ def indices_to_mask(indices, mask_length):
     return mask
+
+
+def flexible_vstack(it, final_len=None):
```
This is not strictly necessary to the feature, and might be too magical. It merely helps users making use of `pairwise_distances_chunked`.

jnothman replied:
I've had a go at removing `flexible_vstack`. In some cases it needs to be replaced by a couple of lines of code, but in radius_neighbors it gets pretty hairy, because it's dealing with arrays of arrays which do not `vstack` nicely.

jnothman committed:
Pass final_len to flexible_vstack

jnothman committed:
Renaming and removing obsolete code

jnothman committed:
Update test_config

jnothman committed:
Fix test import

jnothman committed:
Tweaks

jnothman committed:
Remove debug print

jnothman committed:
Remove unused import

jnothman committed:
Remove flexible_vstack to reduce magic

jnothman committed:
Improve reduce_func description

jnothman committed:
fix flake

jnothman committed:
Block -> chunks; generate_chunks helper

And some cleaning

jnothman committed:
DOC

jnothman committed:
Remove redundant code

jnothman committed:
Document working_memory config

jnothman committed:
Use existing gen_batches

jnothman committed:
TST test_get_chunk_n_rows

jnothman committed:
DOC What's new

jnothman committed:
TST improve pairwise_distances_chunked testing

jnothman committed:
Remove unused import

jnothman committed:
Try fix tests for Python 2

jnothman commented:
This should be ready for review. Marking as MRG.

jnothman committed:
Fix appveyor failure

jnothman committed:
Merge branch 'master' into working_memory

amueller commented on doc/modules/computational_performance.rst:
```diff
@@ -308,6 +308,26 @@ Debian / Ubuntu.
+The following suggests to limit temporary working memory to 128 MiB::
+
+  >>> import sklearn
+  >>> with sklearn.config_context(working_memory=128):
```
This is a good idea, I think. I hadn't imagined this solution, which is probably why I was no super enthusiastic. This doesn't avoid crashes in all cases, though, right? Ideally we'd abort if an algorithm will need more memory. How feasible is that? (I won't have time to give this PR the attention it deserves)

amueller commented:
should this set libsvm's cache?

jnothman commented:
interesting question re libsvm's cache.

I don't think we want to crash if memory demands are too high. we do not
limit absolute memory here. The configuration parameter is more like a
hint. The cases where working_memory can be easily applied are no longer
the ones most at risk of causing a crash.

It would be interesting if numpy had a max_array_size setting beyond which
it raised a warning (which the user could choose to error instead)

jnothman commented:
Currently the default cache size is 200MiB. We can:
1. leave it alone
2. change default `working_memory` to match, and deprecate `cache_size`
3. Benchmark till our heads fall off, and choose a compromise.

amueller commented:
I'm strongly against anything that'll make your head fall of, I think we'd be pretty lost ;)

rth commented on sklearn/utils/__init__.py:
```diff
@@ -506,3 +506,41 @@ def indices_to_mask(indices, mask_length):
+    ----------
+    row_bytes : int
+        The number of bytes consumed by each row of expected output
+        from some function being applied to each chunk.
```
Just reading this docsrting it's unclear what function it could be. Maybe just refer to the array chunk size?

jnothman replied:
I've tried to make this more precise.

rth commented on sklearn/metrics/pairwise.py:
```diff
@@ -1138,6 +1111,118 @@ def _pairwise_callable(X, Y, metric, **kwds):
+    else:
+        if Y is None:
+            Y = X
+        chunk_n_rows = get_chunk_n_rows(row_bytes=8 * _num_samples(Y),
```
`Y` is not square, so I think `row_bytes_` should be `8*Y.shape[1]` not `.shape[0]`. To be more precise couldn't it be something along the lines of,

```py
def _get_row_bytes(X):
    if scipy.sparse.issparse(X):
       # assuming csr
       total_bytes = X.data.nbytes + X.indices.nbytes
    else:
       total_bytes = X.nbytes
    return total_bytes / _num_samples(Y)
```
though this would fail for other input type than dense or csr.

jnothman replied:
I think you've misunderstood.

Each row of a dense (which it should be) distance matrix consists of a float for each row vector in Y.  So the memory we need to account for is len(Y) * 8 bytes per row of X_chunk.

jnothman replied:
The number of features in Y (== in X) may indeed affect temporary memory use, and we could indeed set row_bytes to `8 * (_num_samples(Y) + _num_features(Y))` making some assumptions about how a metric may be implemented. In practice, euclidean has a temporary cost of `X_chunk.shape[0] + Y.shape[0]` for norms; manhattan via scipy does no operations in-place, and so costs `X_chunk.shape[0] * Y.shape[1] * 2` in temporary memory costs. I have now commented in px_dx_chunked that this is not factored in.

If changing the config name to `memory_hint` will help, we can do that...

rth reviewed (commented):
@jnothman A few comments are below

Also, `working_memory` may be a bit misleading, IMO, only a small subset of operations is concerned, maybe `chunk_memory`, or something similar,  would better suggest that this only applies to operations that have been chunked. 

I'll try to look into this in more detail later this week.

jnothman commented:
if we use this for svm cache size, it won't only be about chunked
operations... better name suggestions are welcome.

thanks for the comments!

jnothman commented on sklearn/metrics/cluster/tests/test_unsupervised.py:
```diff
@@ -81,6 +81,49 @@ def test_cluster_size_1():
 
 
+def test_silhouette_paper_example():
+    # Explicitly check per-sample results against Rousseeuw (1987)
```
I'm happy to put it in a separate PR if you wish. Basically, silhouette was not thoroughly enough tested for me to be confident that my changes affected nothing. I'm happy to put it in a separate PR.

jnothman commented on sklearn/_config.py:
```diff
@@ -0,0 +1,90 @@
+        If set, scikit-learn will attempt to limit the size of temporary arrays
+        to this number of MiB (per job when parallelised), often saving both
+        computation time and memory on expensive operations that can be
+        performed in chunks. Global default: 64.
```
I think that's an implementation detail that can be changed over time and application. It will be added as a comment in the `pairwise_distances_chunked` docs.

jnothman commented on sklearn/metrics/pairwise.py:
```diff
@@ -309,16 +316,14 @@ def pairwise_distances_argmin_min(X, Y, axis=1, metric="euclidean",
         Keyword arguments to pass to specified metric function.
 
+    batch_size : integer
+        .. deprecated:: 0.20
```
It's not used in the current code, so description is pointless. If you think that is unwise, I can change it.

jnothman committed:
Respond to Roman

jnothman committed:
Merge branch 'working_memory' of github.com:jnothman/scikit-learn into working_memory

jnothman committed:
See also

rth commented on sklearn/metrics/pairwise.py:
```diff
@@ -1138,6 +1111,127 @@ def _pairwise_callable(X, Y, metric, **kwds):
 
 
+def _check_chunk_size(reduced, chunk_size):
+    is_tuple = isinstance(reduced, tuple)
```
Adding a comment above that ``reduced`` should be either an array/list/sparse matrix or a tuple of arrays/sequence/.. would help comprehension.

jnothman replied:
Do you think we should have this check, or is it unnecessarily constraining?

rth replied:
Well, it's good to have such test, particularly with a user-provided reduce function. Checking each chunk shouldn't be too costly. From my limited understanding of dask, there, they only validate the processing on some sample data (e.g. something the first (few) chunks) then run the full processing without it. But that would add unnecessary complexity here.   Just documenting a bit more this function would help.

rth reviewed (commented):
I have reviewed part of this PR in more detail, comments are below.

It's quite a lot of non-trivial code changes, and I think it might get merged faster if it was split into several smaller PRs (e.g. the context manager mechanism, the `pairwise_distances_chunked` function, and it's applications to silhouette and NN). Though on the other hand, I do understand the advantage of seeing where the more general changes are used in practice. Still currently just reading through the changes and understanding what's happening in detail requires several hours of uninterrupted attention and, generally, I'm not sure how such large PRs are typically reviewed.

rth commented:
> attract other core devs' attention which I feel has been lacking despite the repeated popular interest in these issues!

Just FYI in the case you didn't see the PR, not an actual ping )  @lesteve @ogrisel @qinhanmin2014

jnothman commented:
Thanks for the review. Yes, there are a few things in here, but they are
mostly necessary to illustrate the usefulness and interface for the change.

 The context manager mechanism is in master. I just moved it to a different
file.

jnothman committed:
Merge branch 'master' into working_memory

jnothman commented on doc/modules/computational_performance.rst:
```diff
@@ -308,6 +308,26 @@ Debian / Ubuntu.
+Some calculations when implemented using standard numpy vectorized
+operations involve using a large amount of temporary memory.
+As well as potentially exhausting system memory, memory management
+can overwhelm computation time.  Where computations can be performed
```
I mean swapping, I suppose. We had repeated reports of terribly slow silhouette calculation, and all the time was spent allocating and then filling memory.

rth replied:
Yes, but by chunking NN at least (not sure about `silhouette`) that would still allocate/fill the same amount of memory just sequentially, as far I understand. also related to https://github.com/scikit-learn/scikit-learn/pull/10280#discussion_r160279192

jnothman replied:
Because it does not need to be contiguously addressed and available at the same time it saves swapping time. So you're right, it requires exactly the same number of element accesses, but they can now all be done in main memory where before they had to be swapped to disk.

rth replied:
> but they can now all be done in main memory where before they had to be swapped to disk.

OK, in the context of disk swapping I agree. I just have a swap partition of 0MB which means that anything leading to swapping is treated as a memory error, and I tend to forget that it's not a usual behavior.

jnothman commented on doc/modules/computational_performance.rst:
```diff
@@ -308,6 +308,26 @@ Debian / Ubuntu.
+As well as potentially exhausting system memory, memory management
+can overwhelm computation time.  Where computations can be performed
+in fixed-memory chunks, we attempt to do so, and allow the user to
+hint at the maximum size of this working memory (defaulting to 64 MiB)
```
Probably little effect. I've not done much benchmarking in a while. But basically, it should be a size that is easy for a cheap system to allocate into contiguous main memory.

rth commented:
> Thanks for the review. Yes, there are a few things in here, but they are
mostly necessary to illustrate the usefulness and interface for the change.
>  The context manager mechanism is in master. I just moved it to a different
file.

I agree, was just commenting on the amount of code that needs to reviewed ...

jnothman commented on doc/whats_new/v0.20.rst:
```diff
@@ -130,6 +141,10 @@ Metrics
 
+- :func:`metrics.cluster.silhouette_score` and
+  :func:`metrics.cluster.silhouette_samples` are more memory efficient,
+  causing them to run faster. :issue:`10280` by `Joel Nothman`_.
```
The latter part? silhouette_samples rather than silhouette_score? All the work is done in silhouette_samples... There are some basic benchmarks at https://github.com/scikit-learn/scikit-learn/pull/7177, but I should run them again and more thoroughly.

rth replied:
Yes, thanks for linking the benchmarks. The speedup does look impressive but I'm wondering if the speedup there is also due to some re-write of the code and not just chunking.

rth replied:
I meant about `silhouette_samples` running faster.

jnothman replied:
Again, I should run the benchmarks again, but it *should* be due to the chunking as the bottleneck in the previous implementation was clearly in the pairwise distance calculation.

jnothman commented on sklearn/_config.py:
```diff
@@ -0,0 +1,90 @@
+    """
+    global _ASSUME_FINITE, _WORKING_MEMORY
+    if assume_finite is not None:
+        _ASSUME_FINITE = assume_finite
```
I'm happy to make it a dict. I did initially, but then for a single config it made no sense. If you'd rather...

rth replied:
I'm +1 for a dict.

jnothman committed:
use dict for global config

jnothman commented on sklearn/_config.py:
```diff
@@ -0,0 +1,90 @@
+    -----
+    All settings, not just those presently modified, will be returned to
+    their previous values when the context manager is exited. This is not
+    thread-safe.
```
All settings have a default value, and therefore a previous value.  What it means is that the config state is reset to the state at context entry, not just the state of those settings set by the context manager. I'm not a big fan of this behaviour; I think the settings should be as independent as possible. But it's what we decided to do, following prior art.

jnothman commented on sklearn/metrics/pairwise.py:
```diff
@@ -257,8 +258,14 @@ def euclidean_distances(X, Y=None, Y_norm_squared=None, squared=False,
+
 def pairwise_distances_argmin_min(X, Y, axis=1, metric="euclidean",
-                                  batch_size=500, metric_kwargs=None):
+                                  metric_kwargs=None, batch_size=None):
```
We have previously made changes like this under the tacit assumption that users should be passing all but the first couple of arguments as keywords. I'll revert this reordering.

jnothman replied:
(because in practice I see a lot of use of positional arguments)

rth replied:
Aww, if it's documented, maybe better to change it now indeed than in later versions.

jnothman commented on sklearn/metrics/pairwise.py:
```diff
@@ -1138,6 +1111,127 @@ def _pairwise_callable(X, Y, metric, **kwds):
+        is called repeatedly, where ``D_chunk`` is a contiguous vertical
+        slice of the pairwise distance matrix, starting at row ``start``.
+        It should return an array, a list, or a sparse matrix of length
+        ``D_chunk.shape[0]``, or a tuple of such objects.
```
start parameter is useful when you require a function of position in the array, or have some parallel data you need to extract from. It is thus used in silhouette. Yes, I suppose an example would help.

rth replied:
OK, makes sense, thanks.

jnothman committed:
Minor responses to Roman

jnothman commented:
I've added TODOs for benchmarks and example above.

jnothman committed:
Add pairwise_distances_chunked examples

jnothman committed:
Remove junk in docs

jnothman committed:
Illustrate actual chunking

jnothman committed:
Fix up pairwise_distances_argmin arg ordering

jnothman commented:
One machine's benchmark for `_argmin_min`:
![bench_working_memory](https://user-images.githubusercontent.com/78827/34744108-322b0f60-f5e0-11e7-827c-2649568f1ed4.png)


https://gist.github.com/jnothman/2394b3e199438a500ed9c8f8c9e17500

rth commented:
Interesting, so I imagine a working_memory of 64MB is approximately as good as any other. Thanks for running the benchmarks.

jnothman commented:
for fairness i should probably keep n_samples fixed in X and vary n_samples
only in Y. Lemme try that again.

jnothman commented:
> Interesting, so I imagine a working_memory of 64MB is approximately as good as any other.

Yes I guess that's roughly the point in any case: after a certain amount of memory, runtime becomes constant at best, so you might as well cap the memory and avoid problems directly associated with resources. I'll try to clarify this in the documentation...

jnothman committed:
More nuanced comment on memory-speed tradeoffs

jnothman commented:
We get markedly different results when keeping len(X) fixed and varying len(Y). https://gist.github.com/a88df72e8ec81d778c6a267751b45d7d

![bench_working_memory](https://user-images.githubusercontent.com/78827/34753651-3d3edcec-f60d-11e7-9e94-7ff6e7850fc4.png)

What we see is that len(Y) makes almost no difference. This is probably a good thing. It suggests that `working_memory` is an appropriate parametrization, as opposed to just "number of rows of X per chunk".

On the other hand, we find, for my test system at least, that we get notable speed improvements in runtime for increased working memory until 1GB, which may be a reasonable cap. Indeed we see two near-plateaus (though this may be an artefact of my not having enough samples), perhaps affected by properties of my system.

Certainly it suggests the current proposed default would have slowed down a lot of code. 1024MB might be a closer-to-backwards-compatible-while-fixing-current-issues choice.

This may warrant more benchmarking.

jnothman commented:
Note that n_features affects computation time linearly, but not as a function of working_memory.

jnothman committed:
Increase default working memory to 1GiB

rth commented:
Thanks for the additional benchmarks @jnothman !

To summarize review progress, as things stand now, realistically it doesn't look likely that I'll be able to review fully this PR in the foreseeable future.  It adds 700 LoC and removes 250 LoC, it also references 13 issues. For each of those a reviewer would need through the discussion (and related discussions) and  keep track of possible evolutions (or other approaches to fix the problem) since last review a months ago. This process just does not converge.  

As previously discussed, I understand your motivation to keep it in a single PR. But I still think that something can be done about getting this in faster.

In particular, IMO, it could be the following,
 *  getting a conceptual agreement from few core devs about the general approach taken by this PR
 * submit parts of these modifications as separate PRs (while e.g. we keep this PR whole for illustration)

this would require some more work to sync code between this master PRs and subsequent smaller PRs (I can help with that), but in the end I think it will help to get it merged faster.

@lesteve mentioned a while ago that the ZenHub project management app, has this notion of "Epic" PRs, not sure how this works exactly, but this one is one of those...

jnothman reviewed (commented):
Thanks for the feedback, @rth. I can drop the changes to silhouette and neighbors from this PR, and I shall remove some of the cosmetic things that have snuck in.

jnothman committed:
Undo changes to neighbors and silhouette

Stashing them for later

jnothman committed:
Merge branch 'master' into working_memory

jnothman committed:
Merge branch 'master' into working_memory

jnothman reviewed (commented):
+462 −64 is more acceptable, yes, @rth?

jnothman reviewed (commented):
+462 −64 is more acceptable, yes, @rth?

rth commented:
> +462 −64 is more acceptable, yes,

I confirm it's now possible to scroll through the diff in less than 1 min :)

Thanks @jnothman , I'll try to review it this week.

rth commented on sklearn/metrics/tests/test_pairwise.py:
```diff
@@ -366,10 +371,123 @@ def test_pairwise_distances_argmin_min():
+    assert isinstance(S_chunks, GeneratorType)
+    S_chunks = list(S_chunks)
+    assert len(S_chunks) > 1
+    assert_array_almost_equal(S, np.vstack(S_chunks))
```
numpy recommends to use `assert_allclose` instead https://docs.scipy.org/doc/numpy/reference/generated/numpy.testing.assert_array_almost_equal.html

rth commented on sklearn/metrics/tests/test_pairwise.py:
```diff
@@ -366,10 +371,123 @@ def test_pairwise_distances_argmin_min():
+    assert isinstance(gen, GeneratorType)
+    blockwise_distances = list(gen)
+    min_block_mib = np.array(X).shape[0] * 8 * 2 ** -20
+    working_memory = min(working_memory, min_block_mib)
```
I don't understand why `working_memory` is re-defined after `pairwise_distances_chunked` was called. Essentially we want to check the `len(block)` is either the full array or the amount defined by `working_memory`, right? I think
```py
assert len(block) <= min(..., ...)
``` 
below might be easier to follow.  The `check_pairwise_distances_chunked` is not called with sufficiently small `working_memory` to get 1 row per block, right?

rth commented on sklearn/metrics/tests/test_pairwise.py:
```diff
@@ -366,10 +371,123 @@ def test_pairwise_distances_argmin_min():
+
+    for block in blockwise_distances:
+        memory_used = len(block) * 8
+        assert_true(memory_used <= working_memory * 2 ** 20)
```
`assert memory_user <=..` (or better `len(block) <  ...`).

rth commented on sklearn/utils/__init__.py:
```diff
@@ -506,3 +506,41 @@ def indices_to_mask(indices, mask_length):
+        The maximum return value.
+    working_memory : int or float, optional
+        The number of rows to fit inside this number of MiB will be returned.
+        Defaults to ``sklearn.get_config()['working_memory']``.
```
Same comment as above
> When None the value of sklearn.get_config()['working_memory'] is used which defaults to 1024 MB

rth commented on sklearn/metrics/pairwise.py:
```diff
@@ -1138,6 +1114,176 @@ def _pairwise_callable(X, Y, metric, **kwds):
+        parallel.
+
+        If -1 all CPUs are used. If 1 is given, no parallel computing code is
+        used at all, which is useful for debugging. For n_jobs below -1,
```
Should the `n_jobs` be escaped as code?

rth reviewed (commented):
@jnothman Sorry for the delay in the review. 

A few minor comments are below. Generally, I can't see any other issues with this PR.

jnothman commented:
@rth, no need to apologise. I can even think of a few PRs that have been open for longer ;)

I'll take a look at these another day, though, and my days at the moment are getting pretty cramped.

jnothman commented:
Regarding the use of `assert_allclose`: by default atol=0, which means that comparison to 0 must be absolutely 0. Unfortunately, we know that calculating the distance of some vector to itself is often not identical to 0, for which reason `euclidean_distances` and `cosine_distances` explicitly zero the diagonal when Y is None. As such, my test requires atol=7 to compare the chunked version, which does not explicitly zero the diagonal, to the unchunked version. We could make similar adjustments in `pairwise_distances_chunked`. To do so would assume that the metric was actually a metric, which we do not currently.

jnothman committed:
In response to Roman's comments

TomDLT reviewed (approved):
LGTM

rth commented:
> @TomDLT approved these changes

Great that this is happening!

jnothman committed:
Merge branch 'master' into working_memory

jnothman commented on doc/whats_new/v0.20.rst:
```diff
@@ -78,6 +78,14 @@ Model evaluation
+
+- A new configuration parameter, ``working_memory`` was added to control memory
+  consumption limits in chunked operations, such as the new
+  :func:`metrics.pairwise_distances_chunked`.  See :ref:`working_memory`.
```
I think a reference to the User Guide is most relevant in what's new.

I can add a glossary entry, though I'm not sure how it will help beyond the user guide and the `config_context` docstring.

TomDLT replied:
Fair enough, I just wonder what is the goal of the syntax `:ref:`, which [does not render a link](https://23284-843222-gh.circle-artifacts.com/0/doc/whats_new/v0.20.html#new-features).

Did you mean `:func:set_config`? Or maybe you need a label in `doc/modules/computational_performance.rst`?

jnothman replied:
The latter. A glossary reference would be `:term:`, not `:ref:` which references sections.

jnothman commented on sklearn/metrics/tests/test_pairwise.py:
```diff
@@ -366,10 +372,126 @@ def test_pairwise_distances_argmin_min():
+    min_block_mib = np.array(X).shape[0] * 8 * 2 ** -20
+
+    for block in blockwise_distances:
+        memory_used = len(block) * 8
```
Hmm... indeed!

jnothman commented:
Thanks @TomDLT for the review and the approval! I've addressed all your comments except for the glossary one, which I'm not sure is necessitated at this point.

jnothman committed:
Typo fixes

jnothman committed:
Correct the tests

jnothman commented:
I suppose we could include a new Global Configuration section of the glossary. But I'm not sure how that goes beyond the `config_context` API reference.

jnothman committed:
Add missing ref target

jnothman committed:
Update array formatting in doctest

jnothman committed:
Remove more whitespace in doctest

jnothman committed:
Remove more whitespace in doctest

jnothman commented:
Merging to enable downstream PRs. Let me know if there are further quibbles! Thanks for the reviews, Roman and Tom!

rth commented:
Thanks for this @jnothman !

@TomDLT @amueller FYI there is a follow-up PR in https://github.com/scikit-learn/scikit-learn/pull/11136 that applies this mechanism to brute force nearest neighbors.

rth commented:
.. and also https://github.com/scikit-learn/scikit-learn/pull/11135 for chunking silhouette_score calculations
